{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc3ec7d1",
   "metadata": {},
   "source": [
    "# Creating PCs and GRM\n",
    "\n",
    "Author: Jose Jaime Martinez-Magana\n",
    "\n",
    "Day: 28 March 2023\n",
    "\n",
    "This script was developed to estimate principal component analysis and the GRM using admixture for the Tractor - Mix model for in the Yale HPC McCleary -  cluster. Using the Yale Penn 2 cohort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aa77434",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to the working directory\n",
    "cd /vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix\n",
    "\n",
    "# create the directories to store the results of the pca and grm files\n",
    "mkdir pca_grm\n",
    "\n",
    "# create the directories to store the plink files\n",
    "mkdir plink_files\n",
    "# create the following directories inside the plink file directory\n",
    "mkdir plink_files/prunne_snps plink_files/first_plink plink_files/prunned_plink\n",
    "\n",
    "# create the directories to store the scripts for pca_grm\n",
    "mkdir scripts/pca_grm\n",
    "\n",
    "# move to the directory\n",
    "cd scripts/pca_grm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0f86071",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we will transform the vcf files to plink files and prunned the plink files, named vcfto_plink_prunned.sh\n",
    "\n",
    "# start script\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=yale_penn_vftto_plink_files_prunned\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=18:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1 --cpus-per-task=4\n",
    "#SBATCH --mem-per-cpu=90G\n",
    "#SBATCH --mail-type=ALL\n",
    "####################################################################################\n",
    "# script for building plink files from of Yale Penn 2 cohort vcfs files\n",
    "# day: 28 March 2023\n",
    "# analyzer: Jose Jaime Martinez-Magana - jjm262\n",
    "# cluster: McCleary - HPC Yale\n",
    "####################################################################################\n",
    "# This script uses the Yale Penn cohort filtered by AFR american ancestry higher\n",
    "# than 50%\n",
    "# the data for input to PCAiR\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# set plink files paths\n",
    "yp_pf='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/genotype/phased_vcfs'\n",
    "# set output paths\n",
    "yp_op='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/plink_files'\n",
    "# plink path\n",
    "plink='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/environment/extdata/plink/plink'\n",
    "####################################################################################\n",
    "# run analysis\n",
    "## build plink files for yale penn 2\n",
    "for chr in {1..22}\n",
    "do\n",
    "$plink --vcf \"${yp_pf}\"/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.vcf \\\n",
    "--make-bed \\\n",
    "--keep-allele-order \\\n",
    "--out \"${yp_op}\"/first_plink/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink \\\n",
    "--id-delim '_' \\\n",
    "--threads $SLURM_CPUS_PER_TASK\n",
    "done\n",
    "\n",
    "# after we will prunned the files \n",
    "# running ld prunning\n",
    "for chr in {1..22}\n",
    "do\n",
    "$plink --bfile \"${yp_op}\"/first_plink/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink \\\n",
    "--maf 0.05 \\\n",
    "--hwe 0.000005 \\\n",
    "--indep 50 5 2 \\\n",
    "--threads $SLURM_CPUS_PER_TASK \\\n",
    "--out \"${yp_op}\"/prunne_snps/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink_prune_snps\n",
    "done\n",
    "\n",
    "# removing snps after ld prunning\n",
    "for chr in {1..22}\n",
    "do\n",
    "$plink --bfile \"${yp_op}\"/first_plink/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink \\\n",
    "--extract \"${yp_op}\"/prunne_snps/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink_prune_snps.prune.in \\\n",
    "--make-bed \\\n",
    "--keep-allele-order \\\n",
    "--threads $SLURM_CPUS_PER_TASK \\\n",
    "--out \"${yp_op}\"/prunned_plink/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.plink_prunned_v03282023\n",
    "done\n",
    "# end script\n",
    "# run the previous script with the next command, for this your server should have a slurm handler\n",
    "# if not run your script with your handler\n",
    "sbatch vcfto_plink_prunned.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472f9aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging plink files\n",
    "# requesting computational resources\n",
    "salloc -t 2:00:00 --mem=8G\n",
    "\n",
    "# merging plink files\n",
    "cd /vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/pca_grm\n",
    "yp_op='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/plink_files/prunned_plink'\n",
    "find \"${yp_op}\" | egrep bim | sed 's/.bim//g' > plink_files_merged.list\n",
    "# remove choromosome one manually from plink_files_merged.list\n",
    "# merge with plink\n",
    "# plink path\n",
    "plink='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/environment/extdata/plink/plink'\n",
    "${plink} --bfile \"${yp_op}\"/GWGO_GWCIDR.1kg_phase3_v5.chr1.dose.rsids.filtered.nodup.phased.shapeit.plink_prunned_v03282023 \\\n",
    "--merge-list plink_files_merged.list \\\n",
    "--out \"${yp_op}\"/GWGO_GWCIDR.1kg_phase3_v5.chrmerged.dose.rsids.filtered.nodup.phased.shapeit.plink_prunned_v03292023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36eae24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create directory to store the pca and grm object\n",
    "\n",
    "\n",
    "# creating gds object\n",
    "# script named: create_gds.Rscript\n",
    "\n",
    "#!/usr/bin/env Rscript --vanilla --slave\n",
    "####################################################################################\n",
    "# script for creating PCs with relationship with PCAiR\n",
    "# day: 13 March 2023\n",
    "# author: Jose Jaime Martinez-Magana\n",
    "####################################################################################\n",
    "# This script uses a plink file to create a GDS file for input to PCAiR\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--plinkfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to the directory storing the plink files, without plink extensions *bed/*bim/*fam\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output file name for GDS file, use complete path, example /data/analsyis/analysis.gds\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$plinkfile)){\n",
    "    print_help(opt_parser)\n",
    "    stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "\n",
    "# create list of files\n",
    "bed_f=paste0(opt$plinkfile,\".bed\",sep=\"\")\n",
    "bim_f=paste0(opt$plinkfile,\".bim\",sep=\"\")\n",
    "fam_f=paste0(opt$plinkfile,\".fam\",sep=\"\")\n",
    "\n",
    "# create gds output files\n",
    "snpgdsBED2GDS(bed.fn=bed_f,\n",
    "              bim.fn=bim_f,\n",
    "              fam.fn=fam_f,\n",
    "              family=TRUE,\n",
    "              out.gdsfn=opt$out)\n",
    "# end of script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe10eefa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# script for running the create_gds.Rscript\n",
    "# name of script create_gds.sh\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=yale_penn_build_gds_files\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=10:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1 --cpus-per-task=4\n",
    "#SBATCH --mem-per-cpu=90G\n",
    "#SBATCH --mail-type=ALL\n",
    "####################################################################################\n",
    "# script for constructing GDS files\n",
    "# day: 20 March 2023\n",
    "# analyzer: Jose Jaime Martinez-Magana - jjm262\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "# This script uses the R script to generate GDS files for LD prunned Yale Penn \n",
    "# plink files\n",
    "####################################################################################\n",
    "# load miniconda\n",
    "module load miniconda\n",
    "# activate environment\n",
    "conda activate genesis\n",
    "# run R script for creating GDS files\n",
    "Rscript=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/pca_grm/create_gds.Rscript\"\n",
    "Rscript $Rscript \\\n",
    "--plinkfile=/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/plink_files/prunned_plink/GWGO_GWCIDR.1kg_phase3_v5.chrmerged.dose.rsids.filtered.nodup.phased.shapeit.plink_prunned_v03292023 \\\n",
    "--out=/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/pca_grm/GWGO_GWCIDR.1kg_phase3_v5.chrmerged.dose.rsids.filtered.nodup.phased.shapeit.gds_prunned_v03292023.gds\n",
    "\n",
    "# run script\n",
    "sbatch create_gds.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "348525c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# developing the script for PCA and GRM with PCAiR and PCRelate\n",
    "# name of script: create_grm.Rscript\n",
    "\n",
    "#!/usr/bin/env Rscript --vanilla --slave\n",
    "####################################################################################\n",
    "# R script for creating GRM and PCs\n",
    "# day: 29 March 2023\n",
    "# author: Jose Jaime Martinez-Magana\n",
    "####################################################################################\n",
    "# This script uses a gds file to estimate PCs and GRM using GENESIS\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--gdsfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to gds file\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output path for PCAiR objects, use a complete directory path and a potential name of files. Example: /home/user/out_pca\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$gdsfile)){\n",
    "  print_help(opt_parser)\n",
    "  stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "library(SeqArray)\n",
    "library(parallel)\n",
    "library(BiocParallel)\n",
    "library(stringi);\n",
    "\n",
    "# stablishing a seed\n",
    "set.seed(1000)\n",
    "# stablishing cores\n",
    "cores=detectCores()\n",
    "\n",
    "# create function for LD prunning\n",
    "ld_prun=function(gds){\n",
    "snpset=snpgdsLDpruning(gds,\n",
    "                       method=\"corr\",\n",
    "                       slide.max.bp=10e6,\n",
    "                       ld.threshold=sqrt(0.1),\n",
    "                       maf=0.01,\n",
    "                       missing.rate=0.01,\n",
    "                       verbose=TRUE, num.thread=cores);\n",
    "    pruned=unlist(snpset,\n",
    "                  use.names=FALSE);\n",
    "return(pruned)\n",
    "}\n",
    "\n",
    "\n",
    "# create function for KING-robust analysis\n",
    "king_mat=function(gds){\n",
    "samp.id=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "ibd.robust=snpgdsIBDKING(gds, sample.id=samp.id, family.id=NULL, maf=0.01,missing.rate=0.01,num.thread=cores)\n",
    "return(ibd.robust)\n",
    "}\n",
    "\n",
    "\n",
    "# create function for PCAiR\n",
    "pcair_r=function(gds_geno, pruned, KINGmat){\n",
    "pcair=pcair(gds_geno, snp.include=pruned,\n",
    "            kinobj=KINGmat, divobj=KINGmat)\n",
    "return(pcair)\n",
    "}\n",
    "\n",
    "# create function for PCAiR with sample filter\n",
    "pcair_r_fil=function(gds_geno, pruned, KINGmat, sample_list){\n",
    "    pcair=pcair(gds_geno, snp.include=pruned,\n",
    "                sample.include=sample_list,\n",
    "                kinobj=KINGmat, divobj=KINGmat)\n",
    "                return(pcair)\n",
    "                }\n",
    "\n",
    "\n",
    "# create function to match the sample ID in the GDS with the sample IDs of the supplied sampleID filter\n",
    "filter_samples=function(gds_samples, incl_samples){\n",
    "    matches=unique(grep(paste(incl_samples$SampleID,collapse=\"|\"),\n",
    "                        gds_samples, value=TRUE))\n",
    "    return(matches)\n",
    "}\n",
    "\n",
    "\n",
    "# run analysis for pcs with relationships\n",
    "# open the gds object\n",
    "gds=snpgdsOpen(opt$gdsfile);\n",
    "# LD prunning\n",
    "pruned=ld_prun(gds)\n",
    "# build KING matrix\n",
    "KINGmat=king_mat(gds)\n",
    "# adjust KING matrix\n",
    "KINGmat_m=KINGmat$kinship\n",
    "# add sampleIDs to colnames y row names\n",
    "colnames(KINGmat_m)=KINGmat$sample.id\n",
    "rownames(KINGmat_m)=KINGmat$sample.id\n",
    "# get samples in gds\n",
    "gds_samples=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "# close the gds object\n",
    "snpgdsClose(gds)\n",
    "\n",
    "# read the gds object\n",
    "gds=GdsGenotypeReader(filename=opt$gdsfile)\n",
    "\n",
    "# create a GenotypeData class object\n",
    "gds_geno=GenotypeData(gds)\n",
    "\n",
    "# run PCAiR\n",
    "PCair=pcair_r(gds_geno, pruned, KINGmat_m)\n",
    "PCairpart=pcairPartition(kinobj = KINGmat_m, divobj = KINGmat_m)\n",
    "\n",
    "# PC relate calculation \n",
    "gdsData=GenotypeBlockIterator(gds_geno, snpInclude=pruned)\n",
    "PCrelate=pcrelate(gdsData,\n",
    "                  pcs=PCair$vectors[,1:2],\n",
    "                  training.set=PCairpart$unrels,\n",
    "                  BPPARAM=BiocParallel::SerialParam())\n",
    "\n",
    "# making GRM\n",
    "grm=pcrelateToMatrix(PCrelate)\n",
    "# making sparce matrix\n",
    "grm[grm<0.05]=0\n",
    "grm_sparse=as.matrix(grm, sparse = TRUE)\n",
    "\n",
    "# save file\n",
    "out=list()\n",
    "out$PCair=PCair\n",
    "out$PCairpart=PCairpart\n",
    "out$KINGmat=KINGmat_m\n",
    "out$pruned=pruned\n",
    "out$PCrelate=PCrelate\n",
    "out$grm=grm\n",
    "out$grm_sparse=grm_sparse\n",
    "saveRDS(file=paste0(opt$out,\".rds\",\"\"),out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69026cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# script for running the pca and grm cosntruction\n",
    "# name of script: create_grm.sh\n",
    "\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=yale_penn_build_grm_pcair\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1 --cpus-per-task=8\n",
    "#SBATCH --mem-per-cpu=72G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=day\n",
    "####################################################################################\n",
    "# script for constructing GDS files\n",
    "# day: 29 March 2023\n",
    "# analyzer: Jose Jaime Martinez-Magana - jjm262\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "# This script uses the R script to generate GDS files for LD prunned Yale Penn \n",
    "# plink files\n",
    "####################################################################################\n",
    "# load miniconda\n",
    "module load miniconda\n",
    "# activate environment\n",
    "conda activate tractor_mix\n",
    "# run R script for creating GDS files\n",
    "Rscript=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/pca_grm/create_grm.Rscript\"\n",
    "Rscript $Rscript \\\n",
    "--gdsfile=/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/pca_grm/GWGO_GWCIDR.1kg_phase3_v5.chrmerged.dose.rsids.filtered.nodup.phased.shapeit.gds_prunned_v03292023.gds \\\n",
    "--out=/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/pca_grm/GWGO_GWCIDR.1kg_phase3_v5.chrmerged.dose.rsids.filtered.nodup.phased.shapeit.gds_prunned_grm_pca_v03292023\n",
    "\n",
    "# running script\n",
    "sbatch create_grm.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
