{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc329f69",
   "metadata": {},
   "source": [
    "# Script for running the local ancestry with RFMix\n",
    "\n",
    "Author: Jose Jaime Martinez-Magana\n",
    "\n",
    "Day: 25 February 2023\n",
    "\n",
    "This script was developed to run local ancestry for the Tractor - Mix model for in the Yale HPC - Grace cluster.\n",
    "Using the Yale Penn 2 cohort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b981e605",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to your analysis directory\n",
    "cd palmer_scratch/genomics/yalepenn/tractor_mix/\n",
    "\n",
    "# move to database directory\n",
    "cd databases/\n",
    "\n",
    "# create directory for storing local_ancestry results\n",
    "mkdir local_ancestry\n",
    "# WARNING remember to run mkdir, only once !!!!\n",
    "# if you directories has been created just move to it with cd (Linux)\n",
    "\n",
    "# we will create a sample list for the local ancestry estimation using RFMix\n",
    "# to use only the following populations\n",
    "# FIN Finnish Finnish in Finland\n",
    "# IBS Iberian Iberian populations in Spain\n",
    "# GBR British British in England and Scotland\n",
    "# TSI Toscani Toscani in Italy\n",
    "# YRI Yoruba Yoruba in Ibadan, Nigeria\n",
    "# GWD Gambian Mandinka Gambian in Western Division, The Gambia - Mandinka\n",
    "# MSL Mende Mende in Sierra Leone\n",
    "# ESN Esan Esan in Nigeria\n",
    "# LWK Luhya Luhya in Webuye, Kenya\n",
    "\n",
    "# we have previous downloaded the 1000Genomes database in the following path\n",
    "# /gpfs/gibbs/pi/montalvo-ortiz/reference_panel/1kg/hg37/r20130502\n",
    "# to generate this script we follow the next script in any bash\n",
    "sample_path='/gpfs/gibbs/pi/montalvo-ortiz/reference_panel/1kg/hg37/r20130502'\n",
    "sample_list='integrated_call_samples_v3.20200731.ALL.ped'\n",
    "# set output datapath\n",
    "sample_list_p='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/sample_lists/integrated_call_samples_v3.20200731.ALL.ped.forrfmix'\n",
    "# extract samples\n",
    "cat ${sample_path}/${sample_list} | egrep 'FIN|\\IBS|\\GBR|\\TSI|\\YRI|\\GWD|\\MSL|\\ESN|\\LWK' | egrep 'unrel' | cut -f2,7 | sed 's/FIN/EUR/g' | sed 's/IBS/EUR/g' | sed 's/GBR/EUR/g' | sed 's/TSI/EUR/g' | sed 's/TSI/AFR/g' | sed 's/YRI/AFR/g' | sed 's/GWD/AFR/g' | sed 's/MSL/AFR/g' | sed 's/ESN/AFR/g' | sed 's/LWK/AFR/g' > ${sample_list_p}\n",
    "# add the header manualy to the file\n",
    "vim ${sample_list_p}\n",
    "# IndividualID    Superpopulation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd96236f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding a script for phasing with shapeit2, named local_ancestry_rfmix2.sh\n",
    "# start script\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=yale_penn_local_ancestry_rfmix\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=22:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1 --cpus-per-task=4\n",
    "#SBATCH --mem-per-cpu=350G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=bigmem\n",
    "####################################################################################\n",
    "# script to phased Yale Penn 2 cohort for tractor mix, runned local ancestry\n",
    "# day: 26 February 2023\n",
    "# analyzer: Jose Jaime Martinez-Magana - jjm262\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "# This script uses the Yale Penn cohort filtered information\n",
    "####################################################################################\n",
    "# load conda \n",
    "module load miniconda\n",
    "# activate tractor_mix environment\n",
    "conda activate tractor_mix\n",
    "# set input \n",
    "# set path for input vcf files\n",
    "vcfip='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/genotype/phased'\n",
    "# set path for phased output vcf files\n",
    "vcfop='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/local_ancestry'\n",
    "# set path input for haplotypes of references\n",
    "r_vcf='/gpfs/gibbs/pi/montalvo-ortiz/reference_panel/1kg/hg37/r20130502'\n",
    "# set path input for genetic maps\n",
    "map_ref='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/shapeit/genetic_maps'\n",
    "# set group list, to only include those in the filtering\n",
    "sample_m='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/sample_lists/integrated_call_samples_v3.20200731.ALL.ped.forrfmix'\n",
    "# phasing data with shapeit2\n",
    "for chr in {1..22}\n",
    "do\n",
    "rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr$chr.dose.rsids.filtered.nodup.phased.vcf.gz \\\n",
    "-r ${r_vcf}/ALL.chr$chr.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz \\\n",
    "-m ${sample_m} \\\n",
    "--chromosome=${chr} \\\n",
    "-o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr$chr.dose.rsids.filtered.nodup.phased.localancestry \\\n",
    "-g ${map_ref}/genetic_map_GRCh37_chr$chr.txt \\\n",
    "--n-threads=$SLURM_CPUS_PER_TASK\n",
    "done\n",
    "# end script\n",
    "# run the previous script with the next command, for this your server should have a slurm handler\n",
    "# if not run your script with your handler\n",
    "sbatch local_ancestry_rfmix2.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62658afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# modification day: 27 march 2023\n",
    "# modification for running RFMix in arrays in the servers\n",
    "\n",
    "# making directory to store local ancestry data\n",
    "mkdir /vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/local_ancestry\n",
    "\n",
    "# first we need to create a file with the list of jobs to submit\n",
    "# we generate the content of this file with the following commands\n",
    "vcfip='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/genotype/phased_vcfs'\n",
    "# set path for phased output vcf files\n",
    "vcfop='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/databases/local_ancestry'\n",
    "# set path input for haplotypes of references\n",
    "r_vcf='/gpfs/gibbs/pi/montalvo-ortiz/reference_panel/1kg/hg37/r20130502'\n",
    "# set path input for genetic maps\n",
    "# this map file has to have removed the chr code from all chromosomes\n",
    "map_ref='/gpfs/gibbs/project/montalvo-ortiz/jjm262/analysis/genomic/databases/reference/hapmap/genetic_map/hg37'\n",
    "# set group list, to only include those in the filtering\n",
    "sample_m='/vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/sample_lists/integrated_call_samples_v3.20200731.ALL.ped.forrfmix'\n",
    "\n",
    "# testing script\n",
    "# testing conversion in chromosome 22\n",
    "module load miniconda\n",
    "conda activate tractor_mix\n",
    "# testing conversion\n",
    "rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr3.dose.rsids.filtered.nodup.phased.shapeit.vcf \\\n",
    "-r ${r_vcf}/ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz \\\n",
    "-m ${sample_m} \\\n",
    "--chromosome=3 \\\n",
    "-o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr3.dose.rsids.filtered.nodup.phased.localancestry \\\n",
    "-g ${map_ref}/genetic_map_GRCh37_chr3_red.txt \\\n",
    "--n-threads=36\n",
    "# code runned\n",
    "\n",
    "# move to directory storing the scripts for local ancestry\n",
    "cd /vast/palmer/scratch/montalvo-ortiz/jjm262/genomics/yalepenn/tractor_mix/scripts/local_anc\n",
    "# base code to run\n",
    "rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.vcf -r ${r_vcf}/ALL.chr\"${chr}\".phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz -m ${sample_m} --chromosome=\"${chr}\" -o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.localancestry -g ${map_ref}/genetic_map_GRCh37_chr\"${chr}\"_red.txt --n-threads=$SLURM_CPUS_PER_TASK\n",
    "\n",
    "# creating scripts for chromosomes\n",
    "# chr1 - chr8\n",
    "for chr in {1..8}\n",
    "do \n",
    "echo \"module load miniconda;conda activate tractor_mix;rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.vcf -r ${r_vcf}/ALL.chr\"${chr}\".phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz -m ${sample_m} --chromosome=\"${chr}\" -o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.localancestry -g ${map_ref}/genetic_map_GRCh37_chr\"${chr}\"_red.txt --n-threads=30\"\n",
    "done > job_list_tractormix_rfmix_chr1_chr8_v03272023.txt\n",
    "# chr9 - chr16\n",
    "for chr in {9..16}\n",
    "do \n",
    "echo \"module load miniconda;conda activate tractor_mix;rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.vcf -r ${r_vcf}/ALL.chr\"${chr}\".phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz -m ${sample_m} --chromosome=\"${chr}\" -o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.localancestry -g ${map_ref}/genetic_map_GRCh37_chr\"${chr}\"_red.txt --n-threads=30\"\n",
    "done > job_list_tractormix_rfmix_chr9_chr16_v03272023.txt\n",
    "# chr17 - chr22\n",
    "for chr in {17..22}\n",
    "do \n",
    "echo \"module load miniconda;conda activate tractor_mix;rfmix -f ${vcfip}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.shapeit.vcf -r ${r_vcf}/ALL.chr\"${chr}\".phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz -m ${sample_m} --chromosome=\"${chr}\" -o ${vcfop}/GWGO_GWCIDR.1kg_phase3_v5.chr\"${chr}\".dose.rsids.filtered.nodup.phased.localancestry -g ${map_ref}/genetic_map_GRCh37_chr\"${chr}\"_red.txt --n-threads=30\"\n",
    "done > job_list_tractormix_rfmix_chr17_chr22_v03272023.txt\n",
    "\n",
    "# load module dSQ\n",
    "module load dSQ\n",
    "\n",
    "# generating script for dSQ\n",
    "dsq --job-file job_list_tractormix_rfmix_chr1_chr8_v03272023.txt --mem 500g --time 1-00:00:00 --partition=bigmem --cpus-per-task=30 --mail-type ALL\n",
    "dsq --job-file job_list_tractormix_rfmix_chr9_chr16_v03272023.txt --mem 500g --time 1-00:00:00 --partition=bigmem --cpus-per-task=30 --mail-type ALL\n",
    "dsq --job-file job_list_tractormix_rfmix_chr17_chr22_v03272023.txt --mem 500g --time 1-00:00:00 --partition=bigmem --cpus-per-task=30 --mail-type ALL\n",
    "        \n",
    "# the previous script generates file named dsq-job_list_tractor_phasing_v02282023-2023-02-28.sh\n",
    "sbatch dsq-job_list_tractormix_rfmix_chr1_chr8_v03272023-2023-03-27.sh\n",
    "sbatch dsq-job_list_tractormix_rfmix_chr9_chr16_v03272023-2023-03-27.sh\n",
    "sbatch dsq-job_list_tractormix_rfmix_chr17_chr22_v03272023-2023-03-27.sh\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
